sync/pending_polling.go

```diff
@@ -0,0 +1,471 @@
+package sync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/db"
+	"github.com/NethermindEth/juno/feed"
+	"github.com/NethermindEth/juno/utils"
+)
+
+const (
+	blockHashLag = 10
+)
+
+// needsPreConfirmed reports whether the given protocol version string
+// indicates blocks >= v0.14.0 (i.e., pre_confirmed path is required).
+func needsPreConfirmed(protocolVersion string) (bool, error) {
+	ver, err := core.ParseBlockVersion(protocolVersion)
+	if err != nil {
+		return false, err
+	}
+	return ver.GreaterThanEqual(core.Ver0_14_0), nil
+}
+
+// isGreaterThanTip reports whether the given number is at or beyond the highest known header.
+func (s *Synchronizer) isGreaterThanTip(blockNumber uint64) bool {
+	highest := s.highestBlockHeader.Load()
+	if highest == nil {
+		return false
+	}
+	return highest.Number < blockNumber
+}
+
+// StorePending stores a pending block given that it is for the next height
+func (s *Synchronizer) StorePending(p *core.Pending) (bool, error) {
+	err := blockchain.CheckBlockVersion(p.Block.ProtocolVersion)
+	if err != nil {
+		return false, err
+	}
+
+	head, err := s.blockchain.HeadsHeader()
+	if err != nil {
+		if !errors.Is(err, db.ErrKeyNotFound) {
+			return false, err
+		}
+		head = nil
+	}
+
+	if !p.Validate(head) {
+		return false, fmt.Errorf("store pending: %w", blockchain.ErrParentDoesNotMatchHead)
+	}
+
+	existingPtr := s.pendingData.Load()
+
+	shouldPreserveCurrentPending := existingPtr != nil && *existingPtr != nil &&
+		(*existingPtr).Validate(head) &&
+		(*existingPtr).GetBlock().TransactionCount >= p.Block.TransactionCount
```

Comment:
nit: We can have a separate function, as it looks the same as the logic in preconfirmed. Actually most of the logic are similar too. Can be in a different PR though.